📝(project) add base documentation

jmaupetit opened the pull request:
## Purpose

Now that the project is used in production, we need to explain key concepts and document commands and backends.

## Proposal

- [x] integrate mkdocs
- [x] automate documentation build in the CI
- [x] document backends
- [x] document commands (should be automated given command docstrings)
- [x] ~write tutorial~ postponed (will be handled in another PR)
- [x] improve developer user guide

jmaupetit commented on CHANGELOG.md:
```diff
@@ -42,6 +42,6 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
-[unreleased]: https://github.com/openfun/v1.1.0...master
-[1.1.0]: https://github.com/openfun/ashley/compare/v1.0.0...v1.1.0
-[1.0.0]: https://github.com/openfun/ashley/compare/3d03d85...v1.0.0
```
![](https://media.giphy.com/media/BxWTWalKTUAdq/giphy.gif)

SergioSim commented on .env.dist:
```diff
@@ -22,16 +22,16 @@ RALPH_APP_DIR=/app/.ralph
-# RALPH_SWIFT_OS_PASSWORD=
-# RALPH_SWIFT_OS_REGION_NAME=
-# RALPH_SWIFT_OS_STORAGE_URL=
+# RALPH_SWIFT_OS_AUTH_URL=http://swift:35357/v3/
```
Good Idea to put the test values in the .env.dist, thanks for that)

SergioSim commented on Makefile:
```diff
@@ -93,11 +106,23 @@ logs: ## display app logs (follow mode)
 	@$(COMPOSE) logs -f app
 .PHONY: logs
 
-run: ## starts supported backends servers
```
There are still some references for this command in the Makefile:

At line 65 we have `es-index: run`
At line 136 we have `test: run`

jmaupetit replied:
Good catch! :pray:

SergioSim commented on README.md:
```diff
@@ -1,77 +1,41 @@
-### Local Swift Storage development
-
-To make ralph work with a local instance of Swift:
+> You should sustitute `COMMAND` by the target command, _e.g._ `list`, to see
```
```suggestion
> You should substitute `COMMAND` by the target command, _e.g._ `list`, to see
```

SergioSim commented on docs/backends.md:
```diff
@@ -0,0 +1,133 @@
+
+- `hosts`: a list of cluster hosts to connect to (_e.g._ `["http://elasticsearch-node:9200"]`)
+- `index`: the elasticsearch index where to get/put documents
+- `client_options`: all options from the offical client are supported and can
```
```suggestion
- `client_options`: all options from the official client are supported and can
```

SergioSim commented on docs/contribute.md:
```diff
@@ -0,0 +1,112 @@
+
+Once the project has been bootstrapped, you may want to edit generated `.env`
+file to set up available backends parameters that will be injected in the
+running container as environmnet variables to configure Ralph (see [backends
```
```suggestion
running container as environment variables to configure Ralph (see [backends
```

SergioSim commented on docs/index.md:
```diff
@@ -0,0 +1,75 @@
+# Pull latest docker image and get usage
+$ docker run --rm -i fundocker/ralph:latest ralph --help
+
+# Pro tip: define an alias to ease your life
+$ alias ralph="docker run --rm -i fundocker/ralph:latest ralph"
```
I like this tip :+1:

SergioSim commented on docs/index.md:
```diff
@@ -0,0 +1,75 @@
+decisions.
+
+We try to raise our code quality standards and expect contributors to follow
+the recommandations from our
```
```suggestion
the recommendations from our
```

SergioSim commented on docs/models/index.md:
```diff
@@ -0,0 +1,16 @@
+
+But on the other side, Ralph is a tool dedicated to learning events processing
+so we've implemented key features related to learning events validation and
+convertion. For now, we mostly focus on two learning events standards: [Open
```
```suggestion
conversion. For now, we mostly focus on two learning events standards: [Open
```

SergioSim reviewed (commented):
Beautiful Documentation! Thanks a lot Julien)

jmaupetit commented:
Thank you for your feedback @SergioSim :heart:

quitterie-lcs commented on README.md:
```diff
@@ -1,77 +1,41 @@
+# Ralph
 
-## Getting started with development
+Ralph, is a command-line tool to fetch, extract, convert and push your tracking
```
```suggestion
Ralph is a command-line tool to fetch, extract, convert and push your tracking
```

quitterie-lcs commented on docs/backends.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Backend parameters
+
+To access to OVH's LDP API, you need to register Ralph as an authorized
```
```suggestion
To access OVH's LDP API, you need to register Ralph as an authorized
```

quitterie-lcs commented on docs/contribute.md:
````diff
@@ -0,0 +1,112 @@
+```
+
+Now that you have started at least the `elasticsearch` and `swift` backends,
+it's time to play with those:
````
```suggestion
it's time to play with them:
```

quitterie-lcs commented on docs/index.md:
```diff
@@ -0,0 +1,75 @@
+# Introduction
+
+Ralph, is a command-line tool to fetch, extract, convert and push your tracking
```
```suggestion
Ralph is a command-line tool to fetch, extract, convert and push your tracking
```

quitterie-lcs commented on docs/models/index.md:
```diff
@@ -0,0 +1,16 @@
+edX](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/analytics.html)
+and [xAPI](https://adlnet.gov/projects/xapi/).
+
+Data validation and serialisation/de-serialisation is achieved using
```
```suggestion
Data validation and serialisation/de-serialisation are achieved using
```

quitterie-lcs reviewed (commented):
Congrats for this comprehensive work! 
There is still a very very small issue that comes from the automated update for `command` and `model` documentations: 

- sometimes, there is the description sentence with a point (for example in Open edx events > base > baseContextModel)
 
> Represents the base context model inherited by all event contexts. 

- and sometimes not (for example in ralph > extract)

> Extract input events from a container format using a dedicated parser

quitterie-lcs commented on docs/models/index.md:
```diff
@@ -0,0 +1,16 @@
+
+Data validation and serialisation/de-serialisation is achieved using
+[pydantics](https://pydantic-docs.helpmanual.io) models that are documented in
+the following subsections:
```
```suggestion
the following subsection:
```

quitterie-lcs replied:
As there is, for now, only one bullet point

jmaupetit replied:
Yes. I anticipated a little bit :wink:

jmaupetit commented:
>  There is still a very very small issue that comes from the automated update for command and model documentations

You are totally right, we need to stay consistent. IMO this should be addressed in a new PR.

jmaupetit commented:
@SergioSim & @quitterie-lcs I've addressed all your comments in separated fixup commits. Is there any chance that you have  time for a last review (and validation)? :pray:

quitterie-lcs commented on docs/index.md:
````diff
@@ -0,0 +1,75 @@
+Ralph package can be installed from PyPI using the `pip` tool:
+
+```sh
+# Create a new virtualenv (optionnal)
````
@jmaupetit I just forgot this mistake when reviewing... 
```suggestion
# Create a new virtualenv (optional)
```

quitterie-lcs reviewed (approved):
Let's go! ![](https://media2.giphy.com/media/RrVzUOXldFe8M/giphy.gif)
